proposals/0534-metro-package-exports-support.md

```diff
@@ -0,0 +1,455 @@
+// Loads ./node_modules/pkg/src/utils/y/y.js
+```
+
+Multiple `*`s cannot be matched in patterns ([source](https://github.com/nodejs/node/pull/39635#issue-958542966)):
+
+```js
+"./*/*.js": "./*.js" // Invalid subpath pattern means entry is ignored
+```
+
+**Proposed**:
+
+- P1 (not targeted in first iteration): Metro will resolve subpath patterns.
+
+This feature has a lower prioritisation due to possible implementation complexity and because the introduction of pattern trailers changed its behaviour more recently. The Node.js spec recommends explicitly listed imports for small packages (meaning we anticipate any fixes towards this will be weighted towards large packages).
+
+> For packages with a small number of exports or imports, we recommend explicitly listing each exports subpath entry.\
+— https://nodejs.org/docs/latest-v18.x/api/packages.html#subpath-patterns
+
+### Exact path specifiers
+
+> Package authors should provide either extensioned (`import 'pkg/subpath.js'`) [...] or extensionless (`import 'pkg/subpath'`) [...] subpaths [...] This ensures that there is only one subpath for each exported module so that all dependents import the same consistent specifier.\
+— https://nodejs.org/docs/latest-v18.x/api/packages.html#extensions-in-subpaths
+
+This detail has the potential to conflict with React Native's [platform-specific extensions](https://reactnative.dev/docs/platform-specific-code#platform-specific-extensions), if extensionless subpaths are not offered by the package, under strict handling. Metro's existing resolution behaviour (without `"exports"`) is as follows:
+
+```js
+// For extensions in `resolver.sourceExts`
+
+import FooComponent from './FooComponent';
+// Tries .[platform].js, .native.js, .js (+ TypeScript variants)
+
+import BarComponent from './BarComponent.js';
+// Tries suffixes following the extensioned filename,
+// e.g. .js.[platform].js, .js.native.js, .js.js
+
+// For extensions in `watcher.additionalExts`
+
+import BazComponent from './BazComponent.mjs';
+// Tries exact extension only
+```
+
+The second case above demonstrates less intuitive existing functionality within Metro, which we anticipate has not been in wide use. `"exports"` presents the opportunity to remove this.
+
+**Proposed**:
+
+- Keep current behaviour: Metro will try platform extension variants when resolving both extensionless and extensioned subpaths via `"exports"`.
+- Under strict behaviour, Metro will treat extensioned imports as direct imports and will not attempt to resolve platform extensions.
+- We will communicate to React Native package authors that extensionless exports should be preferred, since they allow the ability to introduce platform-specific modules where needed.
+
+Not committed to in this RFC: We are considering deprecating or changing support for `.native.js` (raised in [facebook/metro#874](https://github.com/facebook/metro/pull/874)). This could be coupled with `"exports"` strict behaviour.
```

Comment:
I hope `.native.js` stays, otherwise I'm not aware of any other way to share a single file between multiple platforms. We have many code bases that share code between native and web. We consistently use `.native.js` for native platforms and `.js` for web.

Reply:
Noted. No near term plans for this change AFAIK — if so it will be stated in a future RFC.